src/passes/GlobalStructInference.cpp

```diff
@@ -0,0 +1,239 @@
+// vtables, and the vtables follow the above pattern, then we can optimize.
+//
+
+#include "ir/find_all.h"
+#include "ir/module-utils.h"
+#include "ir/subtypes.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+struct GlobalStructInference : public Pass {
+  // Maps optimizable struct types to the globals whose init is a struct.new of
+  // them. If a global is not present here, it cannot be optimized.
+  std::unordered_map<HeapType, std::vector<Name>> typeGlobals;
+
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "GlobalStructInference requires nominal typing";
+    }
+
+    // First, find all the information we need. We need to know which struct
+    // types are created in functions, because we will not be able to optimize
+    // those.
+
+    using HeapTypes = std::unordered_set<HeapType>;
+
+    ModuleUtils::ParallelFunctionAnalysis<HeapTypes> analysis(
+      *module, [&](Function* func, HeapTypes& types) {
+        if (func->imported()) {
+          return;
+        }
+
+        for (auto* structNew : FindAll<StructNew>(func->body).list) {
+          auto type = structNew->type;
+          if (type.isRef()) {
+            types.insert(type.getHeapType());
+          }
+        }
+      });
+
+    // We cannot optimize types that appear in a struct.new in a function, which
+    // we just collected and merge now.
+    HeapTypes unoptimizable;
+
+    for (auto& [func, types] : analysis.map) {
+      for (auto type : types) {
+        unoptimizable.insert(type);
+      }
+    }
+
+    // Process the globals.
+    for (auto& global : module->globals) {
+      if (global->imported()) {
+        continue;
+      }
+
+      // We cannot optimize a type that appears in a non-toplevel location in a
+      // global init.
+      for (auto* structNew : FindAll<StructNew>(global->init).list) {
+        auto type = structNew->type;
+        if (type.isRef() && structNew != global->init) {
+          unoptimizable.insert(type.getHeapType());
+        }
+      }
+
+      if (!global->init->type.isRef()) {
+        continue;
+      }
+
+      auto type = global->init->type.getHeapType();
+
+      // We cannot optimize mutable globals.
+      if (global->mutable_) {
+        unoptimizable.insert(type);
+        continue;
+      }
+
+      // Finally, if this is a struct.new then it is one we can optimize; note
+      // it.
+      if (global->init->is<StructNew>()) {
+        typeGlobals[type].push_back(global->name);
+      }
+    }
+
+    // A struct.get might also read from any of the subtypes. As a result, an
+    // unoptimizable type makes all its supertypes unoptimizable as well.
+    // TODO: this could be specific per field (and not all supers have all
+    //       fields)
+    auto unoptimizableCopy = unoptimizable;
+    for (auto type : unoptimizableCopy) {
```

Comment:
Why not iterate over `unoptimizable`? It's not modified in the loop.

Reply:
Good point, done.